docs/src/references/agent/device-profile.md

```diff
@@ -0,0 +1,270 @@
+For example, the firmware module is installed by triggering a `firmware_update` sub-command
+which in turn uses the `firmware_update` workflow for that operation execution.
+Similarly software modules are installed with `software_update` subcommands and
+configuration updates are applied using `config_update` subcommands.
+These subcommands are triggered for each module defined in the profile definition in that order.
+
+Here is a sample device profile workflow:
+
+```toml
+operation = "device_profile"
+
+[init]
+action = "proceed"
+on_success = "scheduled"
+
+# Sort the inputs as desired
+[scheduled]
+script = "/etc/tedge/operations/device_profile.sh ${.payload.status} ${.payload}"
+on_success = "executing"
+on_error = { status = "failed", reason = "fail to sort the profile list"}
+
+[executing]
+operations = [
+    { operation = "firmware_update", target_operation="firmware_update" }
+    { operation = "software_update", target_operation="custom_software_update" }
+    { operation = "config_update" }
+]
+on_exec = "awaiting_sub_operation"
+on_success = "successful"
+on_error = { status = "failed", reason = "fail to apply device profile"}
+
+[awaiting_sub_operation]
+action = "await-operation-completion"
+on_success = "executing"
+on_error = { status = "failed", reason = "fail to update configuration"}
+
+#
+# End states
+#
+[successful]
+action = "cleanup"
+
+[failed]
+action = "cleanup"
+```
+
+* The workflow just proceeds to the `scheduled` state from the `init` state
+* The order of operation execution is finalized in the `scheduled` state.
+  If the `builtin` action is specified in this state, the default order as defined in the input is used.
+  This order can be modified using a `script` action, if desired.
+  The updated list is captured into a `updated_profile` field with the starting `current_index` value.
+  The `current_index` value specifies the next item in the list to be processed by the `executing` state.
+  By default, this `current_index` value starts with `0`.
+  This index value is incremented as the operation execution proceeds.
+  This index can also be used to skip the first `n` items in the list, especially when the same profile is reapplied.
+* The `target_operation` to be invoked for each item in the input is defined in the `executing` state.
+  If a `target_operation` is not explicitly specified, then the default workflow for the `operation` value itself is used.
+  The sub-operation execution starts from the item corresponding to the `current_index` value.
+  As soon as the sub-operation is triggered, the workflow moves to the `awaiting_sub_operation` state defined as the `on_exec` target.
+* In the `awaiting_sub_operation` state, workflow just waits monitoring the state of the sub-operation completion.
+  Once the sub-operation is successful, the workflow moves back to the `executing` state with the index value incremented,
```

Comment:
It might be better to separate the `executing` state from the "processing current item", as the `executing` state is already linked to transitioning the operation state in the cloud, so calling the executing state multiple times might cause some unexpected issues.

Reply:
Split this responsibility into a lightweight `executing` state and a `next_operation` state.